### PR TITLE
Test profiles refactoring

### PR DESCRIPTION
**Pull request by slaskawi**
Hey!

Please take a look at first part of test refactoring.

Highlights:
* There are 2 profiles for running tests - container-tests and remote-tests
* Each container or remote test should have proper c@Category
* Each test type is triggered in it's own phase
* Server is unpacked only once

More comments inlined into the code pieces.

### CODE REVIEW

File: common/pom.xml [hunk 1]
```diff
@@ -57,6 +57,12 @@
       <artifactId>resteasy-jaxrs</artifactId>
     </dependency>
 
+    <dependency>
+      <groupId>org.jboss.pnc</groupId>
```
Comment (slaskawi) [inline common/pom.xml]: This is the only dependency needed for creating standard tests.

File: docker-environment-driver/pom.xml [hunk 2]
```diff
@@ -1,213 +1,138 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-simple</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.jboss.pnc</groupId>
+      <artifactId>test-arquillian-container</artifactId>
```
Comment (slaskawi) [inline docker-environment-driver/pom.xml]: This dependency is needed when creating Arquillian tests which are executed inside a container.

File: integration-test/src/test/java/org/jboss/pnc/integration/deployments/Deployments.java [hunk 3]
```diff
@@ -45,19 +44,19 @@ private static void setTestableWar(EnterpriseArchive ear) {
...
-    }
-
     private static void addEar(Archive<?> webArchive, PomEquippedResolveStage mavenResolver) {
         File[] manuallyAddedLibs = mavenResolver.resolve("org.jboss.pnc:ear-package:ear:1.0-SNAPSHOT").withoutTransitivity().asFile();
         Stream.of(manuallyAddedLibs).forEach(lib -> webArchive.merge(ShrinkWrap.create(ZipImporter.class).
                 importFrom(lib).as(GenericArchive.class)));
     }
+
+    private static void addTestCommonWithTransitives(EnterpriseArchive webArchive, PomEquippedResolveStage mavenResolver) {
+        File[] manuallyAddedLibs = mavenResolver.resolve("org.jboss.pnc:test-common").withTransitivity().asFile();
```
Comment (slaskawi) [inline integration-test/src/test/java/org/jboss/pnc/integration/deployments/Deployments.java]: We need test categories for Arquillian

File: pom.xml [hunk 4]
```diff
@@ -927,6 +952,79 @@
...
+        <configuration>
+          <properties>
+            <property>
+              <name>listener</name>
+              <value>org.jboss.pnc.test.listener.PrintTestNameListener</value>
+            </property>
+          </properties>
+        </configuration>
+        <executions>
+          <execution>
```
Comment (slaskawi) [inline pom.xml]: Each tests category is executed in separate phase - this gives us flexibility (e.g. run unit tests with fork count 2C; remote and container tests without forks).

File: pom.xml [hunk 5]
```diff
@@ -966,34 +1111,75 @@
         <persistence.hibernate.hbm2ddl.auto>create</persistence.hibernate.hbm2ddl.auto>
       </properties>
     </profile>
-    
+
```
Comment (slaskawi) [inline pom.xml]: All profiles which modify properties are placed here. This increases readability.

File: test-arquillian-container/pom.xml [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+                  </artifactItems>
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+
+    <profile>
```
Comment (slaskawi) [inline test-arquillian-container/pom.xml]: All Wildfly/EAP downloading profiles are placed here. We want to download everything only once.

File: test-common/src/main/java/org/jboss/pnc/test/listener/PrintTestNameListener.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.logging.Logger;
+
+public class PrintTestNameListener extends RunListener {
+
+    private Logger logger = Logger.getLogger(PrintTestNameListener.class.getName());
+
+    @Override
+    public void testStarted(Description description) {
+        logger.info("Running test: " + description.getClassName() + "#" + description.getMethodName());
```
Comment (slaskawi) [inline test-common/src/main/java/org/jboss/pnc/test/listener/PrintTestNameListener.java]: Hopefully this will make debugging on Jenkins a bit easier.

**Comment by slaskawi** [comment]
retest this please

**Comment by slaskawi** [comment]
@rnc @matejonnet @mareknovotny @vibe13 @pgier @jdcasey Please take a look at this when you have a while.

**Comment by rnc** [comment]
Looks fine ; talked to @slaskawi offline. 

I think its removed the capability to use a local running Wildfly via `-Darquillian.container.type=remote` (we can always add it back if we want to keep it)

While it might be cleaner to move the test code into appropriate modules this can be done another time. It could well be a fine line on where to put tests in future (depends on whether the test is exercising multiple layers)

**Comment by slaskawi** [comment]
@matejonnet @mareknovotny @vibe13 @pgier @jdcasey last chance, I'll address Nick's comments soon and then I will integrate it

**Comment by slaskawi** [comment]
@rnc Updated, it should be fine now.

**Comment by matejonnet** [comment]
What about adding your pull request line notes to code as comments?

**Comment by slaskawi** [comment]
Added test modules description in `pom.xml`s.

File: docker-environment-driver/src/test/java/org/jboss/pnc/environment/docker/DockerEnvironmentDriverRemoteTest.java [hunk 8]
```diff
@@ -31,22 +15,36 @@
...
+
 /**
  * Unit tests for DockerEnnvironmentDriver
- * 
- * @author Jakub Bartecek <jbartece@redhat.com>
  *
+ * @author Jakub Bartecek <jbartece@redhat.com>
  */
 @RunWith(Arquillian.class)
+@Category({RemoteTest.class, ContainerTest.class})
```
Comment (mareknovotny) [inline docker-environment-driver/src/test/java/org/jboss/pnc/environment/docker/DockerEnvironmentDriverRemoteTest.java]: Why is here 2 categories assigned? why should Remote tests be launched in container testing group?

**Comment by mareknovotny** [comment]
@slaskawi  Are you fixing README.md for running tests too ?

File: docker-environment-driver/src/test/java/org/jboss/pnc/environment/docker/DockerEnvironmentDriverRemoteTest.java (hunk 8 above)
Comment (slaskawi) [inline docker-environment-driver/src/test/java/org/jboss/pnc/environment/docker/DockerEnvironmentDriverRemoteTest.java]: Now you need 2 profiles to run this test - `-Pcontainer-tests` and `-Premote-tests`. Now those 2 groups are separate.

**Comment by slaskawi** [comment]
@mareknovotny Very good point, I'm on it.

**Comment by slaskawi** [comment]
@mareknovotny READ.ME Comment addressed

**Comment by slaskawi** [comment]
@mareknovotny READ.ME Comment addressed

File: docker-environment-driver/src/test/java/org/jboss/pnc/environment/docker/DockerEnvironmentDriverRemoteTest.java (hunk 8 above)
Comment (mareknovotny) [inline docker-environment-driver/src/test/java/org/jboss/pnc/environment/docker/DockerEnvironmentDriverRemoteTest.java]: I think this test is RemoteTest and is launched in  arquillian-weld-se-embedded container, so there will be duplicated container test IMO. WDYT?

Comment (slaskawi) [inline docker-environment-driver/src/test/java/org/jboss/pnc/environment/docker/DockerEnvironmentDriverRemoteTest.java]: In that case I should have used only RemoteTests category. I'll 
introduce this change shortly.

Comment (slaskawi) [inline docker-environment-driver/src/test/java/org/jboss/pnc/environment/docker/DockerEnvironmentDriverRemoteTest.java]: Double checked - this was a container-based test before my changes (see [1]). It should behave exactly the same way after my changes - container-based test which connects to external resources. That's why we need to 2 profiles there.

[1] https://github.com/project-ncl/pnc/pull/243/files#diff-d67f3d8a4636c41bf582c63a0f397bc9L141

**Comment by slaskawi** [comment]
retest this please

**Comment by slaskawi** [comment]
I see some very weird test failures: http://10.3.10.33:8081/job/NCL_SNAPSHOT/229/

Reverting.

**Comment by slaskawi** [comment]
Reverted: https://github.com/project-ncl/pnc/commit/8956db187b6df90302a85a56939429bde32d406e

Will update it tomorrow

**Pull request merged**